File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/02_exchanges_routing/direct_consumer.py
#!/usr/bin/env python3
"""
RabbitMQ Direct Exchange Consumer

Subscribes to specific routing keys (severities).
Can filter to receive only certain types of messages.
"""

import pika
import json
import sys


def create_connection():
    """Create connection to RabbitMQ"""
    credentials = pika.PlainCredentials('admin', 'admin123')
    parameters = pika.ConnectionParameters(
        host='localhost',
        port=5672,
        credentials=credentials
    )
    return pika.BlockingConnection(parameters)


def callback(ch, method, properties, body):
    """Process received messages"""
    try:
        message = json.loads(body.decode('utf-8'))
        severity = method.routing_key

        print(f"\n{'='*60}")
        print(f"📨 Received [{severity.upper()}] message:")
        print(f"   Text: {message['text']}")
        print(f"   Severity: {message['severity']}")
        print(f"   Timestamp: {message['timestamp']}")
        print(f"{'='*60}")

        ch.basic_ack(delivery_tag=method.delivery_tag)
    except Exception as e:
        print(f"✗ Error processing message: {e}")
        ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)


def main():
    """Main function"""
    exchange_name = 'direct_logs'

    # Get severities to subscribe to (default: all)
    severities = sys.argv[1:] if len(sys.argv) > 1 else ['info', 'warning', 'error', 'critical']

    # Connect to RabbitMQ
    connection = create_connection()
    channel = connection.channel()

    # Declare exchange
    channel.exchange_declare(
        exchange=exchange_name,
        exchange_type='direct',
        durable=True
    )

    # Create exclusive queue for this consumer
    result = channel.queue_declare(queue='', exclusive=True)
    queue_name = result.method.queue

    # Bind queue to exchange for each severity
    for severity in severities:
        channel.queue_bind(
            exchange=exchange_name,
            queue=queue_name,
            routing_key=severity
        )
        print(f"✓ Bound to routing key: {severity}")

    # Set up consumer
    channel.basic_qos(prefetch_count=1)
    channel.basic_consume(
        queue=queue_name,
        on_message_callback=callback,
        auto_ack=False
    )

    print(f"\n📡 Waiting for messages with severities: {', '.join(severities)}")
    print("   Press CTRL+C to exit\n")

    try:
        channel.start_consuming()
    except KeyboardInterrupt:
        print('\n✗ Interrupted')
        channel.stop_consuming()
    finally:
        connection.close()


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(f'✗ Error: {e}')
        sys.exit(1)

File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/02_exchanges/direct_consumer.ts
#!/usr/bin/env ts-node
import * as amqp from 'amqplib';
import { Connection, Channel, ConsumeMessage } from 'amqplib';

/**
 * Log message interface
 */
interface LogMessage {
    severity: string;
    message: string;
    timestamp: string;
}

/**
 * Consume messages from direct exchange filtered by severity
 */
async function consumeDirectMessages(severities: string[]): Promise<void> {
    let connection: Connection | null = null;
    let channel: Channel | null = null;

    try {
        const connectionUrl = 'amqp://admin:password@localhost:5672/';
        connection = await amqp.connect(connectionUrl);
        channel = await connection.createChannel();

        // Declare exchange
        const exchangeName = 'logs_direct';
        await channel.assertExchange(exchangeName, 'direct', {
            durable: true
        });

        // Create exclusive queue
        const queueResult = await channel.assertQueue('', { exclusive: true });
        const queueName = queueResult.queue;

        // Bind queue for each severity
        for (const severity of severities) {
            await channel.bindQueue(queueName, exchangeName, severity);
            console.log(`Bound to severity: ${severity}`);
        }

        await channel.prefetch(1);

        console.log(`\n⏳ Waiting for ${severities.join(', ')} messages. Press CTRL+C to exit\n`);

        // Start consuming
        await channel.consume(
            queueName,
            (msg: ConsumeMessage | null) => {
                if (msg !== null) {
                    try {
                        const message: LogMessage = JSON.parse(msg.content.toString());
                        console.log(`✓ [${message.severity}] ${message.message}`);
                        console.log(`  Timestamp: ${message.timestamp}`);
                        channel!.ack(msg);
                    } catch (error) {
                        console.error('Error processing message:', error);
                        channel!.nack(msg, false, false);
                    }
                }
            },
            {
                noAck: false
            }
        );

        // Handle graceful shutdown
        process.on('SIGINT', async () => {
            console.log('\n✓ Consumer stopped');
            try {
                if (channel) await channel.close();
                if (connection) await connection.close();
                process.exit(0);
            } catch (error) {
                console.error('Error during shutdown:', error);
                process.exit(1);
            }
        });

    } catch (error) {
        console.error('Error setting up consumer:', error);

        try {
            if (channel) await channel.close();
            if (connection) await connection.close();
        } catch (closeError) {
            console.error('Error closing connection:', closeError);
        }

        throw error;
    }
}

// Main execution
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.log('Usage: ts-node direct_consumer.ts <severity1> [severity2] ...');
        console.log('Example: ts-node direct_consumer.ts error warning');
        process.exit(1);
    }

    const severities = args;
    consumeDirectMessages(severities).catch(error => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

export { consumeDirectMessages, LogMessage };
